### ISSUE TITLE
The TFLite Converter has errors when "training=True" is used in Dropout and BatchNormalization layers

### ISSUE THREAD

**Issue by EricTheRed20** [tensorflow/tensorflow#50502]
### 1. System information

- OS Platform and Distribution (e.g., Windows 10):
- TensorFlow installation (pip package):
- TensorFlow library (TF 2.5.0):

### 2. Code

When training GANs, some Dropout and BatchNormalization layers need to have "training=True" set for the training to work correctly.  This is as per a [book on GANs](https://machinelearningmastery.com/generative_adversarial_networks/) where the chapter on pix2pix translation is based on [this article](https://phillipi.github.io/pix2pix/).   Unfortunately "training=True" trips up the TensorFlow Lite Converter with the errors shown in the sample code below.  It would be good if the TensorFlow Lite Converter ignored the "training=True" arguments.

The bug can be recreated with the following code.
```
import tensorflow as tf
from keras.models import Input
from keras.models import Model
from keras.layers import Conv2D
from keras.layers import Dropout
from keras.layers import BatchNormalization

# Create minimalist model to show the errors (actual model is much bigger)
in_image = Input(shape=(256,256,3))
out_image = Conv2D(64, (4,4))(in_image)
# In the following two lines, the arguments "training=True" cause errors in the the TensorFlow Lite Converter.
# Removing both "training=True" arguments below makes the conversion work fine, but including the "training=True" causes the errors shown.
out_image = Dropout(0.5)(out_image, training=True)          # Causes error: "TF Select ops: RandomUniform, Details: tf.RandomUniform {device = "", seed = 0 : i64, seed2 = 0 : i64}" below
out_image = BatchNormalization()(out_image, training=True)  # Causes error: "0 incompatible with expected resource" below
model = Model(in_image, out_image)
model.compile()

# Convert model to TensorFlow Lite
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()
with open('model.tflite', 'wb') as f:
  f.write(tflite_model)
```
There might be other layers which can have "training=True" set.

**Comment by karimnosseir** [comment]
For BatchNorm with training=True it requires variables.
You can enable variable support on saved model using

```
converter.experimental_enable_resource_variables = True
```

Flag will be enabled by default later, and currently you can do this only when using from_saved_model - other entry points will be coming soon too.

Note that this will still require TF_Select for the FusedBatchNorm op to run.

For DropOuts:
This requires RandomUniform op, we are working on adding it as a builtin op, for now you can use TF_Select model to run the model.

Thanks

**Comment by EricTheRed20** [comment]
Sorry, but the problem is still there!

After waiting a few days, I then installed the latest nightly build ( tf_nightly-2.7.0.dev20210704-cp37-cp37m-win_amd64.whl ) into a separate virtual environment.

After implementing the changes suggested, the error message when including "training=True" for BatchNormalization() has changed to a different one  (TF Select ops: FusedBatchNormV3
Details: tf.FusedBatchNormV3(tensor<?x253x253x64xf32>, ...  exponential_avg_factor = 0.00999999977 : f32, is_training = true}).  The error message for Dropout() is the same.

The test code now uses "converter.experimental_enable_resource_variables = True" and "from_saved_model()" as specified, and is now as follows:-
```
import tensorflow as tf
from keras.models import Input
from keras.models import Model
from keras.layers import Conv2D
from keras.layers import Dropout
from keras.layers import BatchNormalization

# Create minimalist model to show the errors (actual model is much bigger)
in_image = Input(shape=(256,256,3))
out_image = Conv2D(64, (4,4))(in_image)
# In the following two lines, the arguments "training=True" cause errors in the the TensorFlow Lite Converter.
# Removing both "training=True" arguments below makes the conversion work fine, but including the "training=True" causes the errors shown.
out_image = Dropout(0.5)(out_image, training=True)          # Causes error: "TF Select ops: RandomUniform, Details: tf.RandomUniform {device = "", seed = 0 : i64, seed2 = 0 : i64}" below
out_image = BatchNormalization()(out_image, training=True)  # Causes error: "0 incompatible with expected resource" below
model = Model(in_image, out_image)
model.compile()

# Save the model 
model.save("test_model")

# Convert model to TensorFlow Lite
converter = tf.lite.TFLiteConverter.from_saved_model("test_model")
converter.experimental_enable_resource_variables = True
tflite_model = converter.convert()
with open('test_model.tflite', 'wb') as f:
  f.write(tflite_model)
```
Could someone have a look into it.
Thanks

**Comment by karimnosseir** [comment]
In your sample code, you didn't enable TF Select. You need to enable them since we don't support these ops natively.

See instructions [here](https://www.tensorflow.org/lite/guide/ops_select#convert_a_model)

Basically for conversion you need to add
```
converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
]
```

**Comment by EricTheRed20** [comment]
Thanks, I've added that bit of code.  It looks as though we are going in the right direction but am now getting a different error again "TFLite interpreter needs to link Flex delegate in order to run the model since it contains the following flex op(s): Flex ops: FlexFusedBatchNormV3, FlexRandomUniform".

The code is now as follows:-
```
import tensorflow as tf
from keras.models import Input
from keras.models import Model
from keras.layers import Conv2D
from keras.layers import Dropout
from keras.layers import BatchNormalization

# Create minimalist model to show the errors (actual model is much bigger)
in_image = Input(shape=(256,256,3))
out_image = Conv2D(64, (4,4))(in_image)
# In the following two lines, the arguments "training=True" cause errors in the the TensorFlow Lite Converter.
# Removing both "training=True" arguments below makes the conversion work fine, but including the "training=True" causes the errors shown.
out_image = Dropout(0.5)(out_image, training=True)          # Causes error: "TF Select ops: RandomUniform, Details: tf.RandomUniform {device = "", seed = 0 : i64, seed2 = 0 : i64}" below
out_image = BatchNormalization()(out_image, training=True)  # Causes error: "0 incompatible with expected resource" below
model = Model(in_image, out_image)
model.compile()

# Save the model 
model.save("test_model")

# Convert model to TensorFlow Lite
converter = tf.lite.TFLiteConverter.from_saved_model("test_model")
print("tf.lite.TFLiteConverter.from_saved_model done")
converter.experimental_enable_resource_variables = True
print("experimental_enable_resource_variables done")
converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  tf.lite.OpsSet.SELECT_TF_OPS    # enable TensorFlow ops.
]
print("TF Select done")
tflite_model = converter.convert()
print(Convert done")
with open('test_model.tflite', 'wb') as f:
  f.write(tflite_model)
```

Sorry, but could you have a look at it again.
Thanks
Eric

**Comment by EricTheRed20** [comment]
IMPORTANT UPDATE

In my initial submission I said that "It would be good if the TensorFlow Lite Converter ignored the "training=True" arguments".

On some reflection and investigation, I now realize that the Dropout() and BatchNormalization() layers which are normally not trainable are now trainable.   The training logic for those two layers therefore needs to be implemented in TFLite.  Ignoring "training=True" may prevent an error coming up but won't produce the correct output.
